### PR TITLE
Parallel migration of modules OKAPI-939

### PR DESCRIPTION
**Pull request by adamdickmeiss**
2nd round of change with fixes for optional circulat dependencies.
A real case is mod-source-record-manager + mod-source-record-storage.
mod-source-record-manager uses many interfaces from storage, but
storage optionally depend on optional dependency from manager.

The fix actually reduces code size, because we don't care about
optional dependencies in activation.

### CODE REVIEW

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 1]
```diff
@@ -128,63 +128,59 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
...
-            if (prov.isRegularHandler() && prov.getId().equals(req.getId())) {
-              found = true;
-            }
+    for (InterfaceDescriptor req : md.getRequiresList()) {
+      boolean found = false;
+      for (ModuleDescriptor md1 : modules) {
+        InterfaceDescriptor[] providesList = md1.getProvidesList();
+        for (InterfaceDescriptor prov : providesList) {
+          if (prov.isRegularHandler() && prov.isCompatible(req)) {
+            found = true;
```
[the remark this answers is not in the document]
Comment (steveellis) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:147]: Or perhaps I'm misunderstanding what that outer for loop is doing. In any case the logic is a bit hard to follow.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 2]
```diff
@@ -128,63 +128,59 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
 
   /**
    * Check if md's required and optional interfaces are provided by set of modules.
-   * The interface version is not checked, only the interface name.
-   * Only interfaces listed in allProvided are checked, this allows to exclude some optional
-   * interfaces from the check.
+   * Only interfaces listed in allProvided are checked, this allows for the exclusion of
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:131]: There's no `allProvided` any more.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 3]
```diff
@@ -736,7 +661,58 @@ static void installMaxIterations(
     if (!errors.isEmpty()) {
       throw new OkapiError(ErrorType.USER, String.join(". ", errors));
     }
-    sortTenantModules(tml, modsAvailable, modsEnabled);
+    sortTenantModules(tml, modsAvailable, enabledModules);
+
+  }
+
+  static void sortTenantModules(List<TenantModuleDescriptor> tml,
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:668]: Might be good to have a javadoc here. I'm left wondering what the sort is for. The error message below says its a "topological sort" -- seems important what that is. If it is topological what's the topic, etc.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 4]
```diff
@@ -128,63 +128,59 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
 
   /**
    * Check if md's required and optional interfaces are provided by set of modules.
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:130]: This implementation doesn't check optional interfaces.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 5]
```diff
@@ -784,8 +760,8 @@ static void installMaxIterations(
   /**
    * Find modules that provide a required interface.
    * @param modsAvailable all modules known
-   * @param req required interface to check.
-   * @return newest modules that meets the requirements; empty if no modules are found.
+   * @param req required interface to check
+   * @return newest modules that meets the requirements; empty if no modules are found
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:764]: "modules that that meet the requirements"

File: okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java [hunk 6]
```diff
@@ -103,4 +106,19 @@ public boolean getReinstall() {
   public void setReinstall(boolean reinstall) {
     this.reinstall = reinstall;
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java:107]: `this` can be removed and other setters in the class don't use it.

File: okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java [hunk 7]
```diff
@@ -103,4 +106,19 @@ public boolean getReinstall() {
...
+
+  /**
+   * Set max parallel tenant invocations.
+   * @param maxParallel positive integer
+   */
+  public void setMaxParallel(int maxParallel) {
+    if (maxParallel < 1) {
+      throw new OkapiError(ErrorType.USER, "parallel must be 1 or higher");
+    }
+    this.maxParallel = maxParallel;
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java:122]: Same as above

File: okapi-core/src/main/raml/okapi.raml [hunk 8]
```diff
@@ -46,6 +46,66 @@ traits:
...
+      deploy:
+        description: Whether to deploy (or undeploy if disabling)
+        type: boolean
+        required: false
+        default: false
+      ignoreErrors:
+        description: |
+          Okapi 4.2.0 and later, it is possible to ignore errors during the
+          install operation. This is done by supplying parameter `ignoreErrors=true`.
+          In this case, Okapi will try to upgrade all
```
Comment (steveellis) [inline okapi-core/src/main/raml/okapi.raml:65]: Line breaks could be made more consistent, maybe at 100 chars.

File: okapi-core/src/main/raml/okapi.raml [hunk 9]
```diff
@@ -46,6 +46,66 @@ traits:
...
+        type: boolean
+        required: false
+        default: false
+      simulate:
+        description: Whether the installation is simulated
+        type: boolean
+        required: false
+        default: false
+      tenantParameters:
+        description: Parameters for Tenant init
```
Comment (steveellis) [inline okapi-core/src/main/raml/okapi.raml:106]: "Tenant" probably doesn't need to be capitalized. It isn't in other properties here.

File: okapi-core/src/main/raml/okapi.raml [hunk 10]
```diff
@@ -962,7 +1022,8 @@ traits:
               text/plain:
       post:
         is: [
-          versionFilters
+          versionFilters,
+          installParameters
         ]
         description: Enable, disable or upgrade one or more modules for
           tenant. The request body and response body is of the same type
```
Comment (steveellis) [inline okapi-core/src/main/raml/okapi.raml:1029]: Probably need a period at the end of the sentence here.

File: okapi-core/src/test/java/org/folio/okapi/ProxyTest.java [hunk 11]
```diff
@@ -2010,14 +2010,14 @@ public void testRequestOnly(TestContext context) {
...
         .body(equalTo("[ {" + LS
-            + "  \"id\" : \"request-only-1.0.0\"," + LS
-            + "  \"action\" : \"disable\"" + LS
-            + "}, {" + LS
             + "  \"id\" : \"request-pre-1.0.0\"," + LS
             + "  \"action\" : \"enable\"" + LS
             + "}, {" + LS
             + "  \"id\" : \"auth-f-module-1\"," + LS
             + "  \"action\" : \"enable\"" + LS
+            + "}, {" + LS
```
Comment (steveellis) [inline okapi-core/src/test/java/org/folio/okapi/ProxyTest.java:2018]: Matching like this seems a bit prone to break since it relies on whitespace, line breaks and such being just right (at least if I'm understanding how `equalTo` treats it). Wouldn't it be better to parse the object and check its properties?

But please ignore. I can see this kind of matching is used extensively and I can see how addresses the problem of "I want this complex object with these properties". Anyway just curious about the history of this.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 12]
```diff
@@ -128,63 +128,59 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
...
+        return false;
       }
     }
     return true;
   }
 
-  static void topoSort(List<ModuleDescriptor> modules) {
-    List<ModuleDescriptor> result = new LinkedList<>();
+  /**
+   * Check if md's provided interfaces are required by any of the modules.
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:159]: Check if at least one of md's provided interfaces is a required or optional dependency of any of the modules.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 13]
```diff
@@ -736,7 +661,58 @@ static void installMaxIterations(
...
     }
-    sortTenantModules(tml, modsAvailable, modsEnabled);
+    sortTenantModules(tml, modsAvailable, enabledModules);
+
+  }
+
+  static void sortTenantModules(List<TenantModuleDescriptor> tml,
+      Map<String, ModuleDescriptor> modsAvailable, Collection<ModuleDescriptor> modules) {
+
+    logger.info("sortTenantModules with list {}",
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:671]:
```suggestion
    logger.info("sortTenantModules with list {}", () ->
```

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 14]
```diff
@@ -736,7 +661,58 @@ static void installMaxIterations(
...
+          }
+        }
+      } else {
+        iterator.remove();
+        iterator = tml.iterator();
+        result.add(tm);
+      }
+    }
+    if (!tml.isEmpty()) {
+      throw new OkapiError(ErrorType.USER, "Some modules cannot be topological sorted: "
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:712]: These modules cannot be topological sorted because they have circular or unsatisfiable interface dependencies:

Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:712]: Exactly!

File: okapi-core/src/test/java/org/folio/okapi/InstallTest.java [hunk 15]
```diff
@@ -1514,4 +1519,353 @@ public void installTenantInitVersion2JobError(TestContext context) {
...
+        .header("Content-Type", "application/json")
+        .body(mid.encode()).post("/_/proxy/modules").then().statusCode(201);
+
+    final String tenant = "roskilde";
+    createTenant(context, tenant);
+    deployAsyncInitModule(context, moda.getString("id"), portModule);
+    deployAsyncInitModule(context, modb.getString("id"), portModule);
+    deployAsyncInitModule(context, mid.getString("id"), portModule);
+    ModuleTenantInitAsync tModule = new ModuleTenantInitAsync(vertx, portModule);
+    tModule.start().onComplete(context.asyncAssertSuccess());
```
Comment (julianladisch) [inline okapi-core/src/test/java/org/folio/okapi/InstallTest.java:1564]:
```
    Async start = context.async();
    tModule.start().onComplete(context.asyncAssertSuccess(x -> start.complete()));
    start.await();
```
Otherwise the following post might run before start has completed.

File: okapi-core/src/test/java/org/folio/okapi/InstallTest.java [hunk 16]
```diff
@@ -1514,4 +1519,353 @@ public void installTenantInitVersion2JobError(TestContext context) {
...
+            .add(new JsonObject()
+                .put("id", mid.getString("id"))
+                .put("action", "enable")
+                .put("stage", "done")
+            )
+        );
+    context.assertEquals(jobExpected, job);
+    // moda , modb running in parallel
+    context.assertTrue(tModule.getStartTime(moda.getString("id")).isBefore(tModule.getEndTime(modb.getString("id"))));
+    context.assertTrue(tModule.getStartTime(modb.getString("id")).isBefore(tModule.getEndTime(moda.getString("id"))));
```
Comment (julianladisch) [inline okapi-core/src/test/java/org/folio/okapi/InstallTest.java:1621]: `assertParallelRun(ModuleTenantInitAsync a, ModuleTenantInitAsync b)` and
`assertSequentialRun(ModuleTenantInitAsync a, ModuleTenantInitAsync b)`
can avoid duplicate code.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:147]: Returning true here would only check for one interface (not all).

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 2 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:131]: yep and optional not checked. Fixed.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 17]
```diff
@@ -128,63 +128,59 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
...
-        if (!found) {
-          return false;
-        }
+      }
+      if (!found) {
+        return false;
       }
     }
     return true;
   }
```
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:156]: neat

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 3 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:668]: Agreed. Added some javadoc to the method. This is fun stuff :-)

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 4 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:130]: Correct. Fixed now!

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 5 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:764]: Oh yes.

File: okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java (hunk 6 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java:107]: I don't understand what you mean.

File: okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java (hunk 7 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java:122]: Also here I don't understand.

File: okapi-core/src/main/raml/okapi.raml (hunk 8 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/raml/okapi.raml:65]: Yeah.. It's all over the place.

File: okapi-core/src/main/raml/okapi.raml (hunk 9 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/raml/okapi.raml:106]: sure

File: okapi-core/src/main/raml/okapi.raml (hunk 10 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/raml/okapi.raml:1029]: It's there on next line.. But I can move it up.

File: okapi-core/src/test/java/org/folio/okapi/ProxyTest.java (hunk 11 above)
Comment (adamdickmeiss) [reply okapi-core/src/test/java/org/folio/okapi/ProxyTest.java:2018]: Yes.. Should use JsonObject .. in a lot of places.. Another PR with just that one day.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 12 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:159]: "of any of the" triggers a warning here :-) I understand it.

File: okapi-core/src/test/java/org/folio/okapi/InstallTest.java (hunk 15 above)
Comment (adamdickmeiss) [reply okapi-core/src/test/java/org/folio/okapi/InstallTest.java:1564]: sure

File: okapi-core/src/test/java/org/folio/okapi/InstallTest.java (hunk 16 above)
Comment (adamdickmeiss) [reply okapi-core/src/test/java/org/folio/okapi/InstallTest.java]: ok

File: okapi-core/src/main/raml/okapi.raml [hunk 18]
```diff
@@ -46,6 +46,66 @@ traits:
         enum: ["true","false","only"]
         required: false
         default: "true"
+  installParameters:
+    queryParameters:
+      async:
+        description: Whether to install in the background
```
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply okapi-core/src/main/raml/okapi.raml]: I'll make a separate PR on the RAML

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 19]
```diff
@@ -127,64 +127,55 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
...
+      return false;
     }
     return true;
   }
 
-  static void topoSort(List<ModuleDescriptor> modules) {
-    List<ModuleDescriptor> result = new LinkedList<>();
+  /**
+   * Check if any of md's provided interfaces are used by any of the modules.
+   * <p>A interface is used if it's a required or optional interface dependency.</p>
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:156]: The closing p HTML tag is not required because javadoc uses HTML, not XHTML.
It is very uncommon to use the closing p HTML tag. Therefore some style guides explicitly disallow `</p>`:
https://blog.joda.org/2012/11/javadoc-coding-standards.html
https://metaterasology.github.io/docs/workflow/documentationJavaDoc.html#html-tags
Using `</p>` is uncommon in FOLIO javadoc and Okapi javadoc.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 14 above)
Comment (julianladisch) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:719]: Can you add the full explanation to the error message?

**Review by julianladisch — APPROVED** [review]
Looks good, thanks!

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 19 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java]: ok.

File: okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java (hunk 6 above)
Comment (steveellis) [reply okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java:107]: I didn't notice that the param name was the same as the variable name, so you do need `this`.

File: okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java (hunk 7 above)
Comment (steveellis) [reply okapi-core/src/main/java/org/folio/okapi/util/TenantInstallOptions.java:122]: Same

**Review by steveellis — APPROVED** [review]
